### ISSUE TITLE
iOS can't register onBackgroundEvent or getInitialNotification

### ISSUE THREAD

**Issue by tedi** [notifee/react-native-notifee#119]
We are unable to trigger either the getInitialNotification or onBackgroundEvent events on iOS. Push notifications are coming through fine from FCM on iOS, however the event(s) after a user opens the notification never run. 

This seems to work fine with the "registerNotificationOpened" event from react-native-notifications. The whole reason we purchased Notifee was to remove the need for multiple libraries handling different notification cases, so using both libraries to handle different notification states is not ideal. Any guidance would be appreciated.

Thanks,
Tedi

**Comment by mikehardy** [comment]
Hi there! That doesn't sound like fun. I have these working in my project so I know it's possible, but there may still be a big and/or a documentation problem - hard to say without:

- version information (package.json would be ideal - hopefully up to date)
- the snippet where you register the handlers (are your handlers async? that bites some people and reminds me of this. Are they registered early in startup? Should be before AppRegistry of the app I think...)

Basically the more information you can give the better. Also, it might be useful to do a quick react-native init on a throwaway project just to test this. If you need to start with FCM you can use this https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh to build a throwaway react-native-firebase app then bolt messaging + notifee into it and get a simple App.js reproduction (I hope)

**Comment by tedi** [comment]
Thanks for the quick reply @mikehardy. While I can't share my whole package.json, these are the relevant packages with version numbers:
"react": "16.13.1"
"react-native": "0.63.2",
"@notifee/react-native": "^0.11.1"
"@react-native-firebase/app": "^8.3.0"
"@react-native-firebase/auth": "^8.3.1"
"@react-native-firebase/messaging": "^7.6.1"

I've tried all the methods listed in the documentation, including placing the logic in a async functions and placing the event handler before calling AppRegistry.

**Comment by mikehardy** [comment]
I think the best thing we can do is get a minimal reproducible example going if you can share one, as these APIs appear to work for us. Other possible suspects are splash screens (react-native-splash-screen for instance is a poor choice - it doesn't pass intents correctly on android, vs react-native-bootsplash - although that's not iOS), and I've seen people have difficulty on these handlers with redux-saga before.

If we can show that a minimal app is able to display notifications and handlers are called when notifications receive user interaction you can isolate between library issues or project issues

**Comment by compojoom** [comment]
I was running into nearly the same thing. In our case the getInitialNotification handler seemed to work. What was not working was the `notifee.onBackgroundEvent`. The app is started, I bring it to the background and then I press on the push notification - the app opens at the last screen, but the onBackgroundEvent handler is not being executed. 

This kind of reminded me of this issue:
https://github.com/invertase/react-native-firebase/issues/4104

So I scanned the Notify obj-c code for dispatch and replaced the code in NotifeeApiModule.m
https://github.com/notifee/react-native-notifee/blob/master/ios/RNNotifee/NotifeeApiModule.m#L80

from:
```obj-c
    dispatch_async(dispatch_get_main_queue(), ^{
        if (RCTRunningInAppExtension() || [UIApplication sharedApplication].applicationState == UIApplicationStateBackground) {
            [self sendEventWithName:kReactNativeNotifeeNotificationBackgroundEvent body:eventBody];
        } else {
            [self sendEventWithName:kReactNativeNotifeeNotificationEvent body:eventBody];
        }
    });

```

to:
```obj-c
dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t) (8 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
      if (RCTRunningInAppExtension() || [UIApplication sharedApplication].applicationState == UIApplicationStateBackground) {
          [self sendEventWithName:kReactNativeNotifeeNotificationBackgroundEvent body:eventBody];
      } else {
          [self sendEventWithName:kReactNativeNotifeeNotificationEvent body:eventBody];
      }
    });
```

And then it worked. We'll apparently 8s is too much, but it is just proving that the handler is apparently dispatching before the js side can handle it or something of that sort. Someone who actually knows obj-c maybe can shed some light on this.

@helenaford can you try this change and let us know if then the backgroundHandler is triggered for you?

**Comment by mikehardy** [comment]
Seems like the test case is to programmatically delay app boot on the Javascript side? It should be possible to reproduce this I think in a simple app.js

**Comment by compojoom** [comment]
How do you delay the app boot?
In the case of rn-firebase I was thinking that it has something to do with the loading & parsing of the js bundle. However, here the app is already loaded - it's been sent to the background. When pressing the push notification the app goes in foreground - I don't think that we are reloading the JS here.

**Comment by Salakar** [comment]
Ah delayed dispatch issue again, hello my old friend 🙈 shall we put in a temporary delay 'fix' of say 5s until we have a better solution for RNFB & Notifee on detecting JS bundle readiness?

**Comment by Salakar** [comment]
cc @helenaford 😅 

Relevant context for RNFB: https://github.com/invertase/react-native-firebase/issues/4104

**Comment by compojoom** [comment]
Well, after I posted this I tried it with 1s and it is working great as well. If this fixes it for @tedi I could create PR. But I would really like to understand what the route cause for this is.

**Comment by helenaford** [comment]
Here's a p/r https://github.com/notifee/react-native-notifee/pull/130 with a delay of 1 second delay based on @compojoom's recommendation. @tedi let us know if this would work for you, or it needs to be increased.

**Comment by helenaford** [comment]
release 0.12.2 includes the delay.
